File: sample-app/ios-app/iosApp/Readme.swift
import shared
//import OpenIdConnectClient


struct Readme {
    let client: OpenIdConnectClient
    let tokens: AccessTokenResponse
    let tokenstore: KeychainTokenStore
    let oldAccessToken: String
    
    // Create OpenID config and client:
    func _1() {
        let client = OpenIdConnectClient(
            config: OpenIdConnectClientConfig(
                discoveryUri: "<discovery url>",
                endpoints: Endpoints( // pass nil if you only want to use discovery
                    tokenEndpoint: "<tokenEndpoint>",
                    authorizationEndpoint: "<authorizationEndpoint>",
                    userInfoEndpoint: nil,
                    endSessionEndpoint: "<endSessionEndpoint>",
                    revocationEndpoint: "<revocationEndpoint>"
                ),
                clientId: "<clientId>",
                clientSecret: "<clientSecret>",
                scope: "openid profile",
                codeChallengeMethod: .s256,
                redirectUri: "<redirectUri>",
                postLogoutRedirectUri: "<postLogoutRedirectUri>",
                disableNonce: false
            )
        )
    }
    
    // Create OpenID client with custom http client configuration:
    func _1b() {
        let client = OpenIdConnectClient(
            httpClient: OpenIdConnectClient.companion.DefaultHttpClient.config(block: { config in
                config.installClientPlugin(
                    name: "customheader",
                    onRequest: { requestBuilder, content in
                        requestBuilder.headers.append(name: "User-Agent", value: "oidcclient")
                    },
                    onResponse: {_ in},
                    onClose: {}
                )
            }),
            config: OpenIdConnectClientConfig(
                discoveryUri: "<discovery url>",
                endpoints: nil,
                clientId: "<clientId>",
                clientSecret: "<clientSecret>",
                scope: "openid profile",
                codeChallengeMethod: .s256,
                redirectUri: "<redirectUri>",
                postLogoutRedirectUri: "<postLogoutRedirectUri>",
                disableNonce: false
            )
            
        )
    }
    
    // Request access token using code auth flow:
    func _2() async {
        let factory = CodeAuthFlowFactory(ephemeralBrowserSession: false)
        let flow = factory.createAuthFlow(client: client)
        do {
            let tokens = try await flow.getAccessToken()
        } catch {
            print(error)
        }
    }
    
    // Perform refresh or endSession:
    func _3() async throws {
        try await client.refreshToken(refreshToken: tokens.refresh_token!)
        try await client.endSession(idToken: tokens.id_token!)
    }
    
//     customize endSession request:
    func _3a() async throws {
        try await client.endSession(idToken: "") { requestBuilder in
            requestBuilder.headers.append(name: "X-CUSTOM-HEADER", value: "value")
            requestBuilder.url.parameters.append(name: "custom_parameter", value: "value")
        }
        // endSession with Web flow (opens browser and handles post_logout_redirect_uri redirect)
        let factory = CodeAuthFlowFactory(ephemeralBrowserSession: false)
        let flow = factory.createAuthFlow(client: client)
        try await flow.endSession(idToken: "<idToken>", configureEndSessionUrl: { urlBuilder in
        })
        
    }
    
//     customize getAccessToken request:
    func _3b() async throws {
        let factory = CodeAuthFlowFactory(ephemeralBrowserSession: false)
        let flow = factory.createAuthFlow(client: client)
        try await flow.getAccessToken(
            configureAuthUrl: { urlBuilder in
                urlBuilder.parameters.append(name: "prompt", value: "login")
            },
            configureTokenExchange: { requestBuilder in
                requestBuilder.headers.append(name: "additionalHeaderField", value: "value")
            }
        )
    }
    
    // We provide simple JWT parsing:
    func _4() {
        let jwt = tokens.id_token.map { try! JwtParser.shared.parse(from: $0) }
        print(jwt?.payload.aud) // print audience
        print(jwt?.payload.iss) // print issuer
        print(jwt?.payload.additionalClaims["email"]) // get claim
    }
    
    // TokenStore
    func _5() async throws {
        let tokenstore = KeychainTokenStore()
        try await tokenstore.saveTokens(tokens: tokens)
    }
    
    // RefreshHandler
    func _6() async throws {
        let refreshHandler = TokenRefreshHandler(tokenStore: tokenstore)
        try await refreshHandler.refreshAndSaveToken(client: client, oldAccessToken: oldAccessToken)  // thread-safe refresh and save new tokens to store
    }
}

File: oidc-appsupport/src/androidUnitTest/kotlin/README.kt
import android.os.Bundle
import androidx.activity.ComponentActivity
import io.ktor.client.request.header
import org.publicvalue.multiplatform.oidc.ExperimentalOpenIdConnect
import org.publicvalue.multiplatform.oidc.OpenIdConnectClient
import org.publicvalue.multiplatform.oidc.appsupport.AndroidCodeAuthFlowFactory
import org.publicvalue.multiplatform.oidc.flows.CodeAuthFlowFactory
import org.publicvalue.multiplatform.oidc.tokenstore.TokenRefreshHandler
import org.publicvalue.multiplatform.oidc.tokenstore.TokenStore
import org.publicvalue.multiplatform.oidc.tokenstore.saveTokens
import org.publicvalue.multiplatform.oidc.types.CodeChallengeMethod
import org.publicvalue.multiplatform.oidc.types.parseJwt
import org.publicvalue.multiplatform.oidc.types.remote.AccessTokenResponse
import kotlin.experimental.ExperimentalObjCRefinement


/**
 * Sample code for Readme
 */
@OptIn(ExperimentalObjCRefinement::class, ExperimentalOpenIdConnect::class)
object README {
    val client = OpenIdConnectClient {  }
    val authFlowFactory: CodeAuthFlowFactory = TODO()
    val tokens: AccessTokenResponse = TODO()
    val idToken: String = TODO()
    val tokenstore: TokenStore = TODO()
    val token: String = TODO()
    val refreshHandler: TokenRefreshHandler = TODO()

    // Create OpenID config and client
    fun `Create_OpenID_config_and_client`() {
        val client = OpenIdConnectClient(discoveryUri = "<discovery url>") {
            endpoints {
                tokenEndpoint = "<tokenEndpoint>"
                authorizationEndpoint = "<authorizationEndpoint>"
                userInfoEndpoint = null
                endSessionEndpoint = "<endSessionEndpoint>"
            }

            clientId = "<clientId>"
            clientSecret = "<clientSecret>"
            scope = "openid profile"
            codeChallengeMethod = CodeChallengeMethod.S256
            redirectUri = "<redirectUri>"
            postLogoutRedirectUri = "<postLogoutRedirectUri>"
        }
    }

    // Create AuthFlowFactory in onCreate
    fun `Create_AuthFlowFactory_in_onCreate`() {
        class MainActivity : ComponentActivity() {
            override fun onCreate(savedInstanceState: Bundle?) {
                super.onCreate(savedInstanceState)
                val factory = AndroidCodeAuthFlowFactory()
                factory.registerActivity(this)
            }
        }
    }

    // Request access token using code auth flow
    suspend fun `Request_access_token_using_code_auth_flow`() {
        val flow = authFlowFactory.createAuthFlow(client)
        val tokens = flow.getAccessToken()
    }

    suspend fun `Request_access_token_using_code_auth_flow_continuation`() {
        val flow = authFlowFactory.createAuthFlow(client)
        flow.startLogin()
        if (flow.canContinueLogin()) {
            val tokens = flow.continueLogin()
        }
    }

    // perform refresh or endSession
    suspend fun `perform_refresh_or_endSession`() {
        tokens.refresh_token?.let { client.refreshToken(refreshToken = it) }
        tokens.id_token?.let { client.endSession(idToken = it) }
    }

    // endSession using web flow
    suspend fun `perform_endSession_getrequest`() {
        val flow = authFlowFactory.createEndSessionFlow(client)
        tokens.id_token?.let { flow.endSession(it) }
    }

    // Custom headers/url parameters
    suspend fun `Custom_headers_url_parameters`() {
        client.endSession(idToken = idToken) {
            headers.append("X-CUSTOM-HEADER", "value")
            url.parameters.append("custom_parameter", "value")
        }
    }

    // Custom parameters in authorization request / exchange token request
    suspend fun `Custom auth request`() {
        val flow = authFlowFactory.createAuthFlow(client)
        val tokens = flow.getAccessToken(configureAuthUrl = {
            parameters.append("prompt", "login")
        }, configureTokenExchange = {
            header("additionalHeaderField", "value")
        })
    }

    // We provide simple JWT parsing
    fun `We_provide_simple_JWT_parsing`() {
        val tokens = AccessTokenResponse("abc")
        val jwt = tokens.id_token?.parseJwt()
        println(jwt?.payload?.aud) // print audience
        println(jwt?.payload?.iss) // print issuer
        println(jwt?.payload?.additionalClaims?.get("email")) // get claim
    }


    // token store
    @OptIn(ExperimentalOpenIdConnect::class)
    suspend fun `tokenstore`() {
        tokenstore.saveTokens(tokens)
        val accessToken = tokenstore.getAccessToken()
    }

    // refresh handler
    @OptIn(ExperimentalOpenIdConnect::class)
    suspend fun `refresh_handler`() {
        val refreshHandler = TokenRefreshHandler(tokenStore = tokenstore)
        refreshHandler.refreshAndSaveToken(client, oldAccessToken = token) // thread-safe refresh and save new tokens to store
    }
}
